Convert to `betterproto`

llucax opened the pull request:
This PR migrated the Microgrid API to use `betterproto` instead of `grpc`.

A couple of bugs were found on the way and fixed, one of them directly related to the lack of proper type hints in `grpc`.

The changes are fairly small.

> [!NOTE]
> This PR is just a draft, tests need to be adapted still and I don't plan to merge this to `v0.x.x` directly, as it would be an unnecessary breaking change for the branch used by the SDK. This is mainly to show what changes are needed to migrate a project to `betterproto`.

This PR is using an unreleased repository that hosts the files generated by betterproto from the protobuf files. It will be published soon :tm:

llucax committed:
Update dependencies to use betterproto

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Migrate the client code to use betterproto

Some notable changes:

* The imports can be greatly simplified as now we can import whole
  modules and avoid clashes without having to create one alias per
  symbol.
* We don't need to cast awaitable functions anymore as betterproto has
  the correct type hints
* Because of the above, a bug was uncovered and fixed, where an `await`
  was missing for the API call to `AddInclusionBounds`. I'm not sure if
  this call was not working at all or just errors wouldn't be detected.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Fix call to `list_connections`

Since betterproto use type hints, it expects the `ConnectionFilter`
`starts` and `end` to be `list`s.

We might also just take a `list` in the `connections()` methods instead
to avoid the copy, maybe ensuring that the filter elements are unique
(which was probably the reason to make them a `set`) is not that
important.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

cwasicki commented on pyproject.toml:
```diff
@@ -36,9 +32,10 @@ classifiers = [
+  "betterproto == 2.0.0b6",
   "frequenz-channels >= 1.0.0-rc1, < 2.0.0",
   "frequenz-client-base >= 0.3.0, < 0.4.0",
+  "frequenz-microgrid-betterproto @ file:///home/luca/devel/frequenz-microgrid-betterproto",
```
Is this intended?

cwasicki replied:
>an unreleased repository

Is this meant here?

llucax replied:
Yes, this is just a draft, I didn't crated those repos yet. Will do so soon.

cwasicki reviewed (commented):
Wouldn't the first two commits make sense even when we not migrate to betterproto?

llucax commented:
> Wouldn't the first two commits make sense even when we not migrate to betterproto?

Yes, also fixing the missing `await`, I just left it like that for "presentation purposes", but I think it makes sense to make a separate PR fixing the other issues in this PR that are not related to betterproto :+1:

llucax committed:
Remove unnecessary tests class wrapper

This only adds unnecessary extra indentation.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Stop using the _grpc_server context manager

In future commits we'll remove the testing grpc server in favour of
just mocking the stub, so we don't need to really listen to a port and
include in the unit tests all the gRPC machinery, when we only want to
test the client does what it is supposed to do when getting some data
from the stub.

This also removes another level of extra indentation.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Fix component tests to use betterproto

These tests need just minor changes to the imports.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Fix client test to use betterproto

Fixing this tests is quite complicated because we also stop using a real
gRPC server (`mock_api`) for the tests, to avoid introducing too much
noise and flakiness to *unit tests* (see
https://github.com/frequenz-floss/frequenz-sdk-python/issues/662 for
example).

For now we just tried to replace `mock_api` with `unittest.mock` mocks,
doing the minimal work to get the tests working and move forward with
the migration to betterproto, but in the future we'll provide a higher
level API client mock (see
https://github.com/frequenz-floss/frequenz-client-microgrid-python/issues/39).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove the timeout tests

The timeouts are now tested as part of gRPC failure tests in the client
tests, since timeouts are handled by `grpclib` there isn't much we can
specifically test about timeouts that is not just testing `grpclib`
itself.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove `mock_api` and its tests

We don't use this module anymore and we don't plan to use it for now as
it is very flaky. In the future we might use something similar to do
integration tests.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
This will not pass the tests because it is still waiting for a release of microgrid-betterproto, but it should be ready for the final review, as no code changes should be needed.

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Now everything is green :green_heart:

llucax commented:
Enabling auto-merge.

cwasicki commented on pyproject.toml:
```diff
@@ -2,11 +2,7 @@
-requires = [
-  "setuptools == 68.1.0",
-  "setuptools_scm[toml] == 7.1.0",
-  "frequenz-repo-config[lib] == 0.9.1",
```
Why is repo-config removed here?

llucax replied:
Including `repo-config` as a build dependency in every project was actually a mistake, it was only needed for API projects and now that this is taken care by `setuptools-betterproto` (if we do the switch), it will not be used by any project. My idea is to split repo-config even further in the future, it was an experiment and now that we have a much clear picture about what we need from it, it is a good time to split it into smaller projects and use those smaller projects only where necessary.

llucax replied:
It is unrelated to this PR, though, I just did it now because I realized it was a mistake to include `repo-config` in the build dependencies right after creating `setuptools-betterproto`.

Marenz commented on src/frequenz/client/microgrid/_component_data.py:
```diff
@@ -390,7 +382,7 @@ def from_proto(cls, raw: PbComponentData) -> Self:
         raw_power = raw.inverter.data.ac.power_active
         inverter_data = cls(
             component_id=raw.id,
-            timestamp=raw.ts.ToDatetime(tzinfo=timezone.utc),
```
Hmm we have functions for the conversion in the client library

llucax replied:
This comes from the SDK, predates all the recent work we've been doing with APIs.
